### ISSUE TITLE
CF v237/3232.8 OpenStack DEA Staging Error "cflinuxfs2: No such file or directory"

### ISSUE THREAD

**Issue by gkenaley** [cloudfoundry/cf-release#991]
I have setup my first Cloud Foundry cluster, using 5 OpenStack compute nodes with 1TB RAM and 8TB SSD storage. I am following the CF setup documentation, but have increased the size of the VMs to avoid memory issues. I am using the `cf-rails` and `fib` sample applications, _but have never gotten an application to successfully stage._

I have attached my [latest deployment manifest](https://github.com/cloudfoundry/cf-release/files/334308/cf-deployment.yml.036.txt). It attempts to stay as close as possible to the templates shipped with v237 and the documentation. All SSL is turned off to the degree possible, although I believe I have all the needed keys, certs, and secrets deployed.

I am using a Ubuntu 14.04 PC as a combined Internet gateway, DNS server, and bosh development system. I had to upgrade the Ubuntu PC to Ruby 2.0.0p648 in order to get all of the needed components to install and build. The bind9 DNS server setup sends traffic to the CF-related domains through ha_proxy_z1 floating IP 128.224.250.10:

```
/etc/bind/db.cf-domain.com:cf-domain.com. IN A 128.224.29.34 # Ubuntu PC IP address
/etc/bind/db.system.cf.cf-domain.com:@ IN SOA ubuntu-pc.cf-domain.com. admin ( ...
/etc/bind/db.system.cf.cf-domain.com:  IN NS  ubuntu-pc.cf-domain.com.
/etc/bind/db.system.cf.cf-domain.com  *.system.cf.cf-domain.com. IN	A 128.224.250.10
/etc/bind/db.app.cf.cf-domain.com:    *.app.cf.cf-domain.com.    IN	A 128.224.250.10
/etc/bind/db.cf.cf-domain.com:        *.cf.cf-domain.com.	     IN	A 128.224.250.10
```

There is no load balancer VM or any other physical equipment external to the Openstack system. The external public network 128.224.250.0/24 connects the Ubuntu PC to the OpenStack system. The OpenStack system has an internal private network 192.168.240.0/24 connected to all 18 VMs. In between the public and private network is an OpenStack virtual router with SNAT disabled. The OpenStack security group "cf-bosh" is relaxed well beyond what the documentation for CF and bosh calls for.

The system seems to have the proper buildpacks and stemcell/stack:

```
demo@ubuntu-pc:~/cf-release$ cf buildpacks
Getting buildpacks...
buildpack              position   enabled   locked   filename
staticfile_buildpack   1          true      false    staticfile_buildpack-cached-v1.3.8.zip
ruby_buildpack         2          true      false    ruby_buildpack-cached-v1.6.17.zip
nodejs_buildpack       3          true      false    nodejs_buildpack-cached-v1.5.14.zip
go_buildpack           4          true      false    go_buildpack-cached-v1.7.8.zip
python_buildpack       5          true      false    python_buildpack-cached-v1.5.5.zip
php_buildpack          6          true      false    php_buildpack-cached-v4.3.12.zip
binary_buildpack       7          true      false    binary_buildpack-cached-v1.0.2.zip

demo@gamer-linux:~/cf-release$ cf stacks
Getting stacks in org tis-org / space qa as admin...
OK
name         description
cflinuxfs2   Cloud Foundry Linux-based filesystem
```

To debug this staging problem I turn up the log level on all jobs and then purge all the logs just before deploying. I also monitor the push and staging process from another window using `$ cf logs cf-rails --recent` and see good progress. The push of `cf-rails` proceeds smoothly through over 50 steps and gets to `bundle complete` in 54 seconds.

It finally reaches the tenth write to `/tmp/staged/app` and stops at `Uploading droplet (35M)`:

```
Bundle completed (54.09s)
Cleaning up the bundler cache.
Downloaded [file:///var/vcap/data/dea_next/admin_buildpacks/327ee28d-4b61-4ecc-b7da-b3fa9ba2938d_f5446238d598246474e97f7aea1e8389b26b6a58/dependencies/https___pivotal-buildpacks.s3.amazonaws.com_node_beta-binaries_node-0.12.7-linux-x64.tar.gz]
-----> Preparing app for Rails asset pipeline
Running: rake assets:precompile
    (...snip...)
I, [2016-06-27T04:41:34.827525 #1692]  INFO -- : Writing /tmp/staged/app/public/assets/application-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.css.gz
Asset precompilation completed (4.30s)
Cleaning assets
Running: rake assets:clean
-----> Uploading droplet (35M) [at this instant "$ cf logs cf-rails --recent" immediately shows an error]
```

All apps hang at this point and never fully stage. They have been requested to start, but have not started.

```
$ cf apps
Getting apps in org tis-org / space qa as admin...
OK
name       requested state   instances   memory   disk   urls
cf-rails   started           0/1         512M     1G     cf-rails-meliorative-clamorist.app.cf.cf-domain.com
```

I logged into blobstore_z1 and found that the 35MB droplet had actually been uploaded and was present until the program was terminated.

At that point I bosh sshed to all 17 Vms and grep all the logs for errors. Here are the hits (the other VMs may be unusually quiet):

**ON blobstore_z1 vm: (one error message - looks fatal)**

```
blobstore/internal_access.log:192.168.240.104 - blobstore-username [27/Jun/2016:04:40:30 +0000] "HEAD /admin/system.cf.cf-domain.com-cc-droplets/buildpack_cache/f3/06/f3060397-9ba4-4c4b-b077-9e501ba7a6e8/cflinuxfs2 HTTP/1.1" 404 0 "-" "HTTPClient/1.0 (2.7.1, ruby 2.3.1 (2016-04-26))"

blobstore/internal_error.log:**2016/06/27 04:40:30 [error] 20444#0: *26514 open() "/var/vcap/store/shared/system.cf.cf-domain.com-cc-droplets/buildpack_cache/f3/06/f3060397-9ba4-4c4b-b077-9e501ba7a6e8/cflinuxfs2" failed (2: No such file or directory)**, client: 192.168.240.104, server: blobstore.service.cf.internal, request: "HEAD /admin/system.cf.cf-domain.com-cc-droplets/buildpack_cache/f3/06/f3060397-9ba4-4c4b-b077-9e501ba7a6e8/cflinuxfs2 HTTP/1.1", host: "blobstore.service.cf.internal"
```

**ON runner_z1 VM: (100+ of dea messages/1 warden message)**

```
(100+)dea_next/dea_next.log:{"timestamp":1467003546.6744301,"message":"em-upload.error","log_level":"warn","source":"Staging","data":{"app_guid":"f3060397-9ba4-4c4b-b077-9e501ba7a6e8","task_id":"5f512e35592e4e9d9044a8fdb2c42e7b","destination":"http://cloud-controller-ng.service.cf.internal:9022/staging/droplets/f3060397-9ba4-4c4b-b077-9e501ba7a6e8/upload","connection_count":10,"message":"Error uploading: (Polling failed - status 0; error: Errno::ETIMEDOUT)","http_error":"Errno::ETIMEDOUT","http_status":0,"http_response":""},"thread_id":46918340106640,"fiber_id":46918349548220,"process_id":19230,"file":"/var/vcap/packages/dea_next/lib/dea/utils/upload.rb","lineno":58,"method":"handle_error"}

(1)warden/warden.log: {"timestamp":1467004303.4935195,"message":"Exited with status 0 (1871.971s): [[\"/var/vcap/data/packages/warden/45feaaafb1ed6e77fd1ddd067b2577b8d13c03c7.1-2f5cb2695e36cc8f850e42384684d10854b036b0/warden/src/closefds/closefds\", \"/var/vcap/data/packages/warden/45feaaafb1ed6e77fd1ddd067b2577b8d13c03c7.1-2f5cb2695e36cc8f850e42384684d10854b036b0/warden/src/closefds/closefds\"], \"/var/vcap/data/packages/warden/45feaaafb1ed6e77fd1ddd067b2577b8d13c03c7.1-2f5cb2695e36cc8f850e42384684d10854b036b0/warden/src/oom/oom\", \"/tmp/warden/cgroup/memory/instance-19m567sv61q\"]","log_level":"warn","source":"Warden::Container::Features::MemLimit::OomNotifier","data":{"stdout":"","stderr":""},"thread_id":46982206896520,"fiber_id":46982217237180,"process_id":18887,"file":"/var/vcap/data/packages/warden/45feaaafb1ed6e77fd1ddd067b2577b8d13c03c7.1-2f5cb2695e36cc8f850e42384684d10854b036b0/warden/lib/warden/container/spawn.rb","lineno":135,"method":"set_deferred_success"}
```

**ON doppler_z1 VM: (4 of these messages)**

```
syslog_drain_binder/syslog_drain_binder.stdout.log:{"timestamp":1467002776.742815018,"process_id":4323,"source":"syslog_drain_binder","log_level":"error","message":"Error when polling cloud controller: Get http://api.system.cf.cf-domain.com/v2/syslog_drain_urls?batch_size=1000: dial tcp 128.224.250.10:80: getsockopt: connection timed out","data":null,"file":"/var/vcap/data/compile/syslog_drain_binder/loggregator/src/syslog_drain_binder/main.go","line":87,"method":"main.main"}
```

**ON hm9000_z1 VM: (10 of these messages)**

```
hm9000/hm9000_analyzer.log:{"timestamp":"1467002402.283330202","source":"fetcher","message":"fetcher.HTTP request failed with error","log_level":2,"data":{"error":"Get http://api.system.cf.cf-domain.com/bulk/apps?batch_size=5000\u0026bulk_token={}: dial tcp 128.224.250.10:80: i/o timeout (Client.Timeout exceeded while awaiting headers)"}}
```

**ON loggregator_z1 VM: (5 of these messages)**

```
syslog_drain_binder/syslog_drain_binder.stdout.log:{"timestamp":1467002989.055986404,"process_id":4500,"source":"syslog_drain_binder","log_level":"error","message":"Error when polling cloud controller: Get http://api.system.cf.cf-domain.com/v2/syslog_drain_urls?batch_size=1000: dial tcp 128.224.250.10:80: getsockopt: connection timed out","data":null,"file":"/var/vcap/data/compile/syslog_drain_binder/loggregator/src/syslog_drain_binder/main.go","line":87,"method":"main.main"}
```

So am I correct in assuming that the root cause is the "cflinuxfs2: No such file or directory"?
How would I go about fixing this, and any other problems you might kindly see in these logs?

Thanks,

Gary

**Comment by Amit-PivotalLabs** [comment]
I see a few things.

First of all, the fact that it hangs at "Uploading droplet" jives with the dea_next logs on the runner VMs, where you see the `em-upload.error` and `Errno::ETIMEDOUT`.  The log line is a bit misleading.  It says the `destination` is http://cloud-controller-ng.service.cf.internal:9022/staging/droplets/f3060397-9ba4-4c4b-b077-9e501ba7a6e8/upload.  That's the destination for the initial upload request initiated from DEA to CC, it then polls CC for the status of the upload job, and the URL it hits for that is actually the CC's external URL.  `cloud-controller-ng.service.cf.internal` is the internal, but the URL it ends up polling is the external URL, `api.system.cf.cf-domain.com`.

In some of the other logs (doppler, hm9000, loggregator) we see also see requests to `api.system.cf.cf-domain.com` failing with TCP dial timeouts trying to hit `128.224.250.10:80`.

You should check whether your VMs would have any issue making TCP connections to your floating IP.

**Comment by Amit-PivotalLabs** [comment]
By the way, thanks for all this detailed investigation and providing the information.  One thing that may help (in addition to using the Markdown formatting I've mentioned elsewhere, especially for log or terminal output) is if you can provide links to any documentation or examples you reference.  You've mentioned:

* CF setup documentation
* cf-rails sample application
* fib sample application

Since anyone is free to make documentation and sample applications, it's not necessarily obvious which documentation and sample applications you are actually using.  If you could provide links to the specific things you use, that would be helpful.

**Comment by gkenaley** [comment]
Thanks, Amit.

Our OpenStack system does not by default allow a VM to access its own floating IP.
So ha_proxy could not access its own floating IP 128.224.250.10:80. But all the other VMs should be able to access it. Looks like I need to investigate this thoroughly.
Is there some definitive test I can run?
The following test result looks bad, although the fresh deployment is currently hung forever updating blobstore_z1 after I deleted the old deployment and updated the ha_proxy cert and key:

root@blobstore_z1# curl -k http://192.168.240.125      (private network is okay?)
<html><body><heading1>503 Service Unavailable</heading1>
No server is available to handle this request.
</body></html>

root@blobstore_z1# curl -k http://128.224.250.10        (floating IP blocked, even from other VMs?)
{hangs forever}
^C

The CF setup information I used was almost always on the Cloud Foundry site or on github for release 237 and its sub-projects.

**Docs top level:** [http://docs.cloudfoundry.org/](url)
**Installation**: [http://docs.cloudfoundry.org/deploying/openstack/index.html](url)
**Release Notes:** [https://github.com/cloudfoundry/cf-release/releases](url)
**Components:** [https://github.com/cloudfoundry/](url)

Finding proper sample applications is extremely difficult, I have tried to write various simple programs but there is always some missing or improper manifest, requirements.txt, setup.py, buildpack, or obsolete parameters. I get close with Ruby and python, but Java and node.js have fatal setup problems. Most of the other samples require access to 3rd-party services. I'd settle for hello world looping for 5 minutes right now.

**Sample Apps I usually try:**
**cf-rails**: [https://github.com/cloudfoundry-samples/cf-sample-app-rails](url)
**fib**: [https://github.com/cloudfoundry-samples/fib-cpu/](url)
**rabbitstack**: [http://rabbitstack.github.io/openstack/cloud%20foundry/devops/deploying-cloud-foundry-on-openstack-juno-and-xenserver-part-ii/](url)

I have also tried java and python simple looping programs.

Thanks,

Gary

**Comment by Amit-PivotalLabs** [comment]
I'm not a networking expert, so I'm not sure the best way to fix your problem, but you'll definitely need to be able to make TCP connections from your internal VMs to your floating IP.  I'm not familiar with OpenStack but I'm more familiar with AWS.  The issue you're seeing seems similar to things I've seen on AWS when you have a misbehaving NAT which is used between your private and public subnets.  You mentioned "in between the public and private network is an OpenStack virtual router with SNAT disabled."  You may need to look into configuring that router correctly.

As for sample applications, you can look at the ones [here](https://github.com/cloudfoundry/cf-acceptance-tests/tree/master/assets) used for acceptance testing Cloud Foundry.  If this is not a production Cloud Foundry deployment, you could even consider running these acceptance tests against your environment ([see README](https://github.com/cloudfoundry/cf-acceptance-tests)).  I would fix the networking issue before even attempting to run acceptance tests though.

**Comment by gkenaley** [comment]
I will try to develop a workaround for the floating IP access problem.

This is the problem using a single OpenStack virtual router, according to a networking expert:

"The issue here is that you have 2 different VM instances behind the same router and instead of using their private addresses (since they are on the same network) they are trying to use their floating IP addresses.  Because of how floating IP addresses are implemented that will not work unless the VM instances are behind 2 different routers.  When the VM instances are behind the same router it prevents one FIP instancs from seeing the other’s ARP requests (because they are on the same switch port)."

By the way, I cntrl-C'd that hung blobstore_z1 update and re-started "bosh deploy" - this time it finished.
I am now able to login with a self-signed certificate for "*.system.cf.cf-domain.com" for the first time:

$ cf login -a **https://api.system.cf.cf-domain.com** -u admin -p admin
API endpoint: **https://api.system.cf.cf-domain.com**
Authenticating...
**OK**
Targeted org tis-org
API endpoint:   https://api.system.cf.cf-domain.com (API version: 2.56.0)
User:           admin
Org:            tis-org
Space:          No space targeted, use 'cf target -s SPACE'

Thanks,

Gary

**Comment by gkenaley** [comment]
I tried this temporary workaround: executing the following command on all 16 CF VMs, the bosh director VM, and the Ubuntu host PC. It maps the ha_proxy floating IP back to its private IP for this lab test setup.
[https://ask.openstack.org/en/question/454/cannot-reach-floating-ip-from-vm/](url)

**echo '192.168.240.125 128.224.250.10' >> /etc/hosts**

On the next "cf push" after clearing the logs, the error messages regarding floating IPs seem to be gone from the logs mentioned above, but the other errors seem to remain. I will be looking at the logs in more detail later today and will post what I find.

**Comment by Amit-PivotalLabs** [comment]
@gkenaley were you able to make any progress on this?

**Comment by gkenaley** [comment]
Hi Amit,

No, I was not able to make any progress. I tried to cross-check with another version of OpenStack, but it was so difficult to install that I could not finish the deployment.

(1) Are we 100% convinced that inability to access the ha_proxy floating IP from the VM is the issue?

(2) Is there some other way to configure the system so that the VM access of the ha_proxy floating IP is not necessary?

(3) Would you be interested in remotely accessing the system to research the issue?

Thanks,

Gary

**Comment by Amit-PivotalLabs** [comment]
(1) Yes, I'm fairly convinced that this is the issue.

(2) There are several components that need to talk to other components in Cloud Foundry.  One way this is done in some cases is to have a target component register a route with the gorouter, and the source component know about this route and make requests to that route.  The DNS resolution of those routes will typically resolve to the HA Proxy external IP (floating IP), and so in that case the source component will need to be able to reach the HA Proxy external IP (from there, the HA Proxy and gorouter act as proxies, sending the request to the actual target component).

One thing you could try is running a little DNS server behind the firewall that resolves your system and apps domains to the HA Proxy internal IP, and then update the networks in your manifest to point to your internal DNS server first so that it can resolve the address, before recursing to an external DNS server which will inevitable resolves those domains to the public IP.

**Comment by gkenaley** [comment]
I control two DNS (bind9) servers - one on the Ubuntu PC jump node (128.224.29.34) that implements the public network (128.224.250.0/24) via VLAN), so I can set them up as you describe. It may take a week or two to get it done, as I have to rebuild the system, and vacations are in the way.

**Zone Files on DNS server 128.224.29.34:** # Point to ha_proxy_z1 private network address:
db.app.cf.cf-domain.com: *.app.cf.cf-domain.com. IN A 128.224.240.125
db.cf.cf-domain.com: *.cf.cf-domain.com. IN A 128.224.240.125
db.system.cf.cf-domain.com: *.system.cf.cf-domain.com. IN A 128.224.240.125

forwarders: 128.224.29.100  # Points to floating address if recursed to

**cf-manifest.yml:**
 dns:
    - 128.224.29.34  # Recursion enabled on Ubuntu DNS server; forwards to 128.224.29.100

Thanks,

Gary

**Comment by gkenaley** [comment]
**Success!**
I can now run the example apps fib and cf-rails for the first time.

First I changed the zone files to point to ha_proxy_z1's private address, as described in the previous post, but the system just hung.

Finally I commented out the 9 lines in the deployment manifest which setup the public network and the floating IP address for ha_proxy_z1. So now the ha_proxy has only a private address and I can run apps on an OpenStack system without hairpin SNAT, using Neutron networking.

**[cf-deployment.yml.043.txt](https://github.com/cloudfoundry/cf-release/files/362920/cf-deployment.yml.043.txt)**

This final change to the deployment manifest only re-launched the ha_proxy_z1 VM, and I see these non-fatal errors when launching each of these 2 apps:
[STG/0]      ERR 
[STG/0]      OUT -----> Uploading droplet (35M)
[DEA/0]      OUT Starting app instance (index 0) with guid d06cab7c-9dda-4696-a551-ad141de7b567
[App/0]      ERR [2016-07-14 00:19:47] INFO  WEBrick 1.3.1
[App/0]      ERR [2016-07-14 00:19:47] INFO  ruby 2.3.0 (2015-12-25) [x86_64-linux]
[App/0]      ERR [2016-07-14 00:19:47] INFO  WEBrick::HTTPServer#start: pid=29 port=61003

I tried to run the following very simple python program, but it crashes. I thought stdout gets posted to the logs:
$ cat pycount.py 

import time
count = 0
while count < 300:
   print (count, " is  less than 300")
   count = count + 1
   time.sleep(1)
else:
   print (count, " is not less than 300")

Not sure what the problem is.

Thanks,

Gary
